src/Types.elm

```diff
@@ -2,18 +2,73 @@ module Types exposing (..)
 
 type Tree a = Node a (List (Tree a))
 
-type Token = TTSC TokTSC | TokAssign | TokLParen | TokRParen | TokHole | TokVar String
-             | TokConstInt Int | TokConstBool Bool | TokInvalid | TokBackSlash | TokArrow | TokEnd
+type Token = TTSC TokTSC | TokAssign | TokLParen | TokRParen | TokHole |
+              TokVar String | TokHasType | TokConstInt Int | TokConstBool Bool |
+              TokInvalid | TokBackSlash | TokArrow | TokEnd | TokTypeName String
 
 type TokTSC = TTSCInt TokTSCInt | TTSCBool TokTSCBool
 type TokTSCInt = TokPlus | TokMinus | TokTimes
 type TokTSCBool = TokEq | TokAnd | TokOr
 
 type Const = CBool Bool | CInt Int
 type Term = CTerm Const | VTerm String | Plus Term Term | Minus Term Term | Times Term Term
-            | Eq Term Term | And Term Term | Or Term Term | Lam Term Term | App Term Term
+            | Eq Term Term | And Term Term | Or Term Term | Lam String Term | App Term Term
             | MissingInt | MissingBool | Missing | EmptyTree
 
+-- V(alue)Type is a type that a TreeAssembly term can evaluate to
+{-
+  TAny may be used when the type can be either int or bool (discuss)
+  TNone is used as a default value for types we haven't determined yet (similar to EmptyTree)
+-}
+type VType = TBool | TInt | TFun VType VType | TAny | TNone
+
 type alias Var =
   { name: String
-  , term: Term }
+  , term: Term
+  , vtype: VType}
+
+{-
+  Returns the type signature of basic language constructs (if it's not relevant, return TNone)
+-}
+getTypeSignature : Term -> VType
+getTypeSignature t =
+  case t of
+    CTerm (CBool b) -> TBool
+    CTerm (CInt i) -> TInt
+    Plus x y -> TFun (TFun TInt TInt) TInt
+    Minus x y -> TFun (TFun TInt TInt) TInt
+    Times x y -> TFun (TFun TInt TInt) TInt
+    Eq x y -> TFun (TFun TInt TInt) TBool
+    And x y -> TFun (TFun TBool TBool) TBool
+    Or x y -> TFun (TFun TBool TBool) TBool
```

Comment:
This definition of Lam's type doesn't quite seem right to me.  For instance, if we use this function on the term `\x.x+x`, it looks like it would get the type of `x+x`, which is `TFun (TFun TInt TInt) TInt`, but the lambda should have the type `TFun TInt TInt`, since it's a function from int to int.

Reply:
This function is currently only used for predefined functions (binary operations). It is not used for getting the types of lambdas. Currently, type checking for lambdas is assumed to be correct. For instance:
`a :: Int -> Int
 a = \x -> x + 2`
would have the type
`Int -> Int`
because the user said it had that type. This is a temporary feature that will be addressed once we have type inference going.

Reply:
Ah gotcha!  Sounds like you've got the right then.